### PR TITLE
Update Derivative for PositionPoint objects

### LINKED ISSUE #4849: [0.9.5-rc3] (regression) Derivative does not work with MIN/MAX/LAST
Hey guys,
Looks like something has changed since 0.9.4. Derivative no longer works with MIN/MAX/LAST aggregate functions... Here is some sample data and queries:

```
SELECT * FROM test WHERE time > now() - 1h

time					value
2015-11-19T21:00:54Z	942562514911
2015-11-19T21:05:53Z	942858150171
2015-11-19T21:11:00Z	943171809495
2015-11-19T21:15:55Z	943444939393
2015-11-19T21:20:58Z	943706044968
2015-11-19T21:26:00Z	943940460815
2015-11-19T21:30:58Z	944141789469
2015-11-19T21:35:57Z	944382269230
2015-11-19T21:40:57Z	944603490366
2015-11-19T21:45:58Z	944773315039
2015-11-19T21:50:55Z	944985627521
2015-11-19T21:56:00Z	946093669442

SELECT DERIVATIVE(SUM(value), 1s) FROM "test" WHERE time > now() - 1h GROUP BY time(5m)

time					value
2015-11-19T21:00:00Z	657522.08
2015-11-19T21:05:00Z	985450.8666666667
2015-11-19T21:10:00Z	1045531.08
2015-11-19T21:15:00Z	910432.9933333333
2015-11-19T21:20:00Z	870351.9166666666
2015-11-19T21:25:00Z	781386.1566666666
2015-11-19T21:30:00Z	671095.5133333333
2015-11-19T21:35:00Z	801599.2033333334
2015-11-19T21:40:00Z	737403.7866666666
2015-11-19T21:45:00Z	566082.2433333333
2015-11-19T21:50:00Z	707708.2733333333
2015-11-19T21:55:00Z	3693473.07

SELECT DERIVATIVE(LAST(value), 1s) FROM "test" WHERE time > now() - 1h GROUP BY time(5m)

time					value
2015-11-19T21:00:00Z
2015-11-19T21:05:00Z	
2015-11-19T21:10:00Z	
2015-11-19T21:15:00Z	
2015-11-19T21:20:00Z	
2015-11-19T21:25:00Z	
2015-11-19T21:30:00Z	
2015-11-19T21:35:00Z	
2015-11-19T21:40:00Z	
2015-11-19T21:45:00Z	
2015-11-19T21:50:00Z	
2015-11-19T21:55:00Z	
```

This is obviously an issue because we don't always want to average the values, sometimes we only want the last value reported for that bucket, as is the case with this data. This data is coming from SNMP counter values, which means they are always increasing sequentially. If for some reason we stored the data twice inside of the same 5 minute interval (maybe a lag in the collection agent, etc), we only really care about the last value that was inserted. If we were to use SUM/MEAN or any other aggregate function, it will distort the data.

This was working on previous versions of 0.9.x...

### PR DESCRIPTION
**Pull request by nathanielc**
The selectors that return PositionPoint were broken. Added  logic to convert PositionPoint objects to float64. This fixes #4849. 

I added integration tests for derivative functions with all other valid functions.

One variation doesn't work and I am curious if my test is correct or not.
A count query with fill previous doesn't work as expected since a group by interval with no points has a well defined count of 0. So the fill logic never sees a nil value to fill in with the previous value. 

Give a query: `SELECT derivative(count(value)) from db0.rp0.cpu where time >= '2010-07-01 18:47:00' and time <= '2010-07-01 18:47:03' group by time(2s) fill(previous)`

With a dataset that is missing a all data points for a 2s interval the count returns 0 so the fill(previous) doesn't work and the derivative goes negative.

@corylanou @pauldix Thoughts on whether count with group by should return 0 or nil?

### CODE REVIEW

File: cmd/influxd/run/server_test.go [hunk 1]
```diff
@@ -1501,6 +1501,403 @@ func TestServer_Query_SelectRawCalculus(t *testing.T) {
...
+		&Query{
+			name:    "calculate derivative of count with unit default (2s) group by time with fill previous",
+			command: `SELECT derivative(count(value)) from db0.rp0.cpu where time >= '2010-07-01 18:47:00' and time <= '2010-07-01 18:47:03' group by time(2s) fill(previous)`,
+			exp:     `{"results":[{"series":[{"name":"cpu","columns":["time","derivative"],"values":[["2010-07-01T18:47:02Z",0]]}]}]}`,
+		},
+		&Query{
+			name:    "calculate derivative of count with unit 4s group by time with fill previous",
+			command: `SELECT derivative(count(value), 4s) from db0.rp0.cpu where time >= '2010-07-01 18:47:00' and time <= '2010-07-01 18:47:03' group by time(2s) fill(previous)`,
+			exp:     `{"results":[{"series":[{"name":"cpu","columns":["time","derivative"],"values":[["2010-07-01T18:47:02Z",0]]}]}]}`,
+		},
```
Comment (nathanielc) [inline cmd/influxd/run/server_test.go]: This is the failing count + derivative test.

**Comment by pauldix** [comment]
looks good. Count should return 0. I think either fill shouldn't work on count, or fill should fill in any zero values on count. Nil and 0 are the same with a count query. What you think?

**Comment by nathanielc** [comment]
I think fill should fill in any zeros on Count. I'll see how easy that is to get working.

**Comment by corylanou** [comment]
I vote that count returns nil (or nothing) like all other aggregates do, and if you want it `filled`, you use `fill` like all other aggregates. This keeps everything consistent, plus makes it very easy for derivative to do the correct thing as well.

**Comment by pauldix** [comment]
Count returning nil was actually a bug that we recently fixed. For other aggregates it makes sense to return nil, but not with count. Count is always defined. If there are no data points, it's zero.

**Comment by otoolep** [comment]
https://github.com/influxdb/influxdb/pull/4778/files

File: tsdb/aggregate.go [hunk 2]
```diff
@@ -342,11 +342,12 @@ func (e *AggregateExecutor) processFill(results [][]interface{}) [][]interface{}
...
 	}
 
+	isCount := e.stmt.HasSimpleCount()
 	// They're either filling with previous values or a specific number
 	for i, vals := range results {
 		// start at 1 because the first value is always time
 		for j := 1; j < len(vals); j++ {
-			if vals[j] == nil {
+			if vals[j] == nil || (isCount && isZero(vals[j])) {
 				switch e.stmt.Fill {
```
Comment (nathanielc) [inline tsdb/aggregate.go]: Here is the logic for filling zero count values.

@corylanou @pauldix Thoughts?

File: cmd/influxd/run/server_test.go (hunk 1 above)
Comment (nathanielc) [inline cmd/influxd/run/server_test.go]: With my latest commit this is now passing, by explicitly using fill for count 0 time buckets.

File: tsdb/aggregate.go [hunk 3]
```diff
@@ -342,11 +342,12 @@ func (e *AggregateExecutor) processFill(results [][]interface{}) [][]interface{}
 		return newResults
 	}
 
+	isCount := e.stmt.HasSimpleCount()
 	// They're either filling with previous values or a specific number
 	for i, vals := range results {
 		// start at 1 because the first value is always time
 		for j := 1; j < len(vals); j++ {
-			if vals[j] == nil {
+			if vals[j] == nil || (isCount && isZero(vals[j])) {
```
Comment (corylanou) [inline tsdb/aggregate.go]: Makes sense.

**Comment by corylanou** [comment]
Generally looks good.  I would like to see a test with `count(distinct value)` in a derivative call and see it fail to make sure the check is working properly.  Otherwise +1.

**Comment by nathanielc** [comment]
@corylanou not sure I follow which test case you want. Something like this:

```
SELECT derivative(count(distinct(value))) FROM xxxx fill(previous)
```

Fails with an error about the query not containing a field before any of the fill logic can be evaluated. The code doesn't understand triply nested functions as far as I can tell.

File: cmd/influxd/run/server_test.go [hunk 4]
```diff
@@ -1501,6 +1501,408 @@ func TestServer_Query_SelectRawCalculus(t *testing.T) {
...
+		&Query{
+			name:    "calculate derivative of count of distinct with unit default (4s) group by time with fill previous",
+			command: `SELECT derivative(count(distinct(value))) from db0.rp0.position where time >= '2010-07-01 18:47:00' and time <= '2010-07-01 18:47:07' group by time(4s) fill(previous)`,
+			exp:     `{"results":[{"error":"aggregate call didn't contain a field derivative(count(distinct(value)))"}]}`,
+		},
```
Comment (nathanielc) [inline cmd/influxd/run/server_test.go]: Triple nested function fail check.

**Comment by corylanou** [comment]
:shipit:

**Pull request merged**